File: QuixBugs-fixed/Python/wrap.py
def wrap(text, cols):
    lines = []
    while len(text) > cols:
        # Try to find a space to break the line, prioritizing breaking at spaces
        end = text.rfind(' ', 0, cols + 1)
        if end == -1:  # If no space found, force break at cols
            end = cols
        line, text = text[:end], text[end:]
        lines.append(line)
    lines.append(text)  # Append remaining text as a line
    return lines
'''
The original code failed to correctly handle the case where the remaining text was shorter than or equal to the column width \text{(cols)}, leading to an infinite loop or incorrect output. Additionally, it incorrectly handled spaces when breaking lines. By checking if the length of the remaining text is less than or equal to \text{cols}, we can append the rest of the text to the lines list and break the loop. The adjustment of the \text{end} variable (+1) ensures that spaces at the end of a line are correctly included in the line, rather than being the first character of the next line. This approach ensures that all words are kept whole unless they exceed the column width, and lines are filled as much as possible up to the specified width. Tracking the length of the remaining text (\text{len(text)}) and the position to break the line (\text{end}) was crucial in identifying and fixing the issue.

'''

File: QuixBugs-fixed/Java/wrap.java
import java.util.*;
public
class WRAP {
public
  static void main(String[] args) {
    System.out.println("abc".lastIndexOf("c", 30));
  }
public
  static ArrayList<String> wrap(String text, int cols) {
    ArrayList<String> lines = new ArrayList<String>();
    String line;
    while (text.length() > cols) {
      int end = text.lastIndexOf(" ", cols);
      if (end == -1) {
        end = cols - 1;
      }
      if(end == 0){
          end = cols;
      }
      line = text.substring(0, end);
      text = text.substring(end);
      lines.add(line);
    }
    if(text.length()!=0){
      lines.add(text);
    }
    return lines;
  }
}
/* Tracking parameters `text` and `end` reveals that when `end` equals 0, `line` would be a single space, and the next `line` would also have a single space as its first character, since `text = text.substring(end);`. This behavior can be observed in the failed test cases: the output lines can only be wrapped at spaces, resulting in no more than a single space at the start of each line. 
Thus, checking if `end` equals 0 enables handling the scenario when the text is longer than the column width and does not have a space within `cols` to wrap at. In this case, the program wraps at `end = cols`.
 */